### PR TITLE
Task99/esp from vcf

### PR DESCRIPTION
**Pull request by at7**
Required updates to allow reading ESP allele and genotype frequencies and counts from VCF. [jira ticket](https://www.ebi.ac.uk/panda/jira/browse/ENSVAR-2377)

### CODE REVIEW

**Comment by at7** [comment]
Working on my [sandbox](http://ves-hx2-76.ebi.ac.uk:5040/Homo_sapiens/Variation/Population?db=core;r=1:230709548-230710548;v=rs699;vdb=variation;vf=19167044) I have highlighted ESP data from VCF as such. But this will not go live like this.

**Comment by helensch** [comment]
In test_data_ESP_GRCh37.vcf.gz, one ID has '~' in front of rs. Is this expected?
17	3417996	~rs72314855

**Comment by at7** [comment]
The example is taken from the original ESP file. The id doesn't matter

File: modules/t/VCFCollection.t [hunk 1]
```diff
@@ -167,7 +168,54 @@ $coll->use_seq_region_synonyms(1);
 $gts = $coll->get_all_SampleGenotypeFeatures_by_VariationFeature($vf);
 ok($gts && scalar @$gts == 3, "get_all_SampleGenotypeFeatures_by_VariationFeature with synonyms 3");
 
+ok($coll->vcf_collection_close, 'close VCF collection filehandle');
+
+## test ESP data query
+$v = $va->fetch_by_name('rs80359165');
+($vf) = @{$v->get_all_VariationFeatures};
+ok($vf && $vf->isa('Bio::EnsEMBL::Variation::VariationFeature'), "get variation feature for rs32954037");
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/t/VCFCollection.t]: There is no entry in variation_feature.txt for rs32954037

[the remark this answers is not in the document]
Comment (helensch) [reply modules/t/VCFCollection.t]: The code says fetch_by_name('rs80359165')
The comment says get variation feature for rs32954037
Was querying why they were different

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm [hunk 2]
```diff
@@ -777,6 +779,140 @@ sub get_all_Alleles_by_VariationFeature {
...
+    if (defined $info_gts) {
+      my @gts = split(',', $info_gts);
+      foreach my $gt (@gts) {
+        my @tmp_gt = ();
+        if ($gt =~ /^[ACGT]+$/i) {
+          my @gt_alleles = split('', $gt);
+          foreach my $gt_allele (@gt_alleles) {
+            push @tmp_gt, $gt_allele;
+          }
+        } elsif ($gt =~ /[A\d|R]+/) { #Observed Genotypes. For INDELs, A1, A2, or An refers to the N-th alternate allele while R refers to the reference allele.
```
Comment (helensch) [inline modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: Suggest adding test for INDELs. The code is called in test in populationGenotypeAdaptor.t

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm [hunk 3]
```diff
@@ -777,6 +779,140 @@ sub get_all_Alleles_by_VariationFeature {
...
+              my $idx = shift @gt_variables;
+              if (my $allele = $allele_map{$idx - 1}) {
+                push @tmp_gt, $allele;
+              } else {
+                my $allele = $vcf_alts->[$idx - 1];
+                push @tmp_gt, $allele;
+              }
+            }  
+            else {
+              warn "Unexpected genotype variable. Expected values are R, A1, A2, An"
```
Comment (helensch) [inline modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: Suggest adding test for warnings

File: modules/Bio/EnsEMBL/Variation/DBSQL/PopulationGenotypeAdaptor.pm [hunk 4]
```diff
@@ -273,24 +273,43 @@ sub fetch_all_by_Variation {
...
+    if($use_vcf) {
+      my $vfs = $variation->get_all_VariationFeatures;
+
+      if($vfs && @$vfs) {
+        @from_vcf =
+          map {$_->{adaptor} = $self; $_}
+          map {@{$_->get_all_PopulationGenotypes_by_VariationFeature($vfs->[0], $population)}}
+          @{$self->db->get_VCFCollectionAdaptor->fetch_all() || []};
+      }
+      push @$cached, @from_vcf;
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/DBSQL/PopulationGenotypeAdaptor.pm]: moved into if block

File: modules/Bio/EnsEMBL/Variation/DBSQL/PopulationGenotypeAdaptor.pm [hunk 5]
```diff
@@ -273,24 +273,43 @@ sub fetch_all_by_Variation {
...
+
+      # Add a constraint on the variation_id column and pass to generic fetch
+      my $constraint = qq{ pg.variation_id = $variation_id };
     
-    # If required, add a constraint on the population id
-    if (defined($population)) {
-      my $population_id = $population->dbID();
-      $constraint .= qq{ AND pg.population_id = $population_id };
+      # If required, add a constraint on the population id
+      if (defined($population)) {
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/DBSQL/PopulationGenotypeAdaptor.pm]: added test for passing population object and actually fixed a bug in the PopulationGenotypeAdaptor

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm [hunk 6]
```diff
@@ -777,6 +779,140 @@ sub get_all_Alleles_by_VariationFeature {
...
+
+  my @pops = @{$self->get_all_Populations};
+  @pops = grep {$_->name eq $given_pop->name || ($_->{_raw_name} || '') eq $given_pop->name} @pops if $given_pop;
+
+  my $variation = $vf->variation;
+  my $population_genotype_adaptor = $self->use_db ? $self->adaptor->db->get_PopulationGenotypeAdaptor : undef;
+
+  my @population_genotypes = ();
+
+  while(1) {
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: last unless $vcf->{record} and  last unless @$matched should prevent running into an infinite loop. Can you can think of a case where it wouldn't break out of the while loop?

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm [hunk 7]
```diff
@@ -777,6 +779,140 @@ sub get_all_Alleles_by_VariationFeature {
...
+  }
+  # get data from VF
+  my $vf_ref = $vf->ref_allele_string;
+  my $vf_alts = $vf->alt_alleles;
+  # check initial VCF match
+  my $matched = $self->_get_matched_alleles_VF_VCF($vf, $vcf);
+  return [] unless @$matched;
+
+  my @pops = @{$self->get_all_Populations};
+  @pops = grep {$_->name eq $given_pop->name || ($_->{_raw_name} || '') eq $given_pop->name} @pops if $given_pop;
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: done

File: modules/t/VCFCollection.t [hunk 8]
```diff
@@ -167,7 +168,54 @@ $coll->use_seq_region_synonyms(1);
...
+    'p:ESP6500:EA a:C f:0.0001 c:1'
+  ],
+  'get_all_Alleles_by_VariationFeature - freqs and counts ESP rs80359165'
+);
+
+my @population_genotypes = @{$coll->get_all_PopulationGenotypes_by_VariationFeature($vf)};
+
+is_deeply(
+  [
+    map {'p:'.$_->population->name.' genotype:'.$_->genotype_string.' f:'.sprintf("%.4f", $_->frequency).' c:'.$_->count}
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/t/VCFCollection.t]: fixed

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm [hunk 9]
```diff
@@ -633,83 +633,87 @@ sub get_all_Alleles_by_VariationFeature {
...
+            if defined($ref_freq_index) || scalar @split > scalar @$vcf_alts;
+
+          for my $i(0..$#split) {
+            my $f = $split[$i];
+            next if $f eq '.';
+            $total_afs->{$pop_id} += $f;
+
+            if(my $allele = $allele_map{$i}) {
+              $freqs->{$pop_id}->{$allele} = $f;
+              $counts->{$pop_id}->{$allele} = sprintf('%.0f', $f * $an) if $an;
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: This has been like this for a while and is not related to the PR. If you want to further investigate if we are doing the right thing here could you please put in a ticket? Thank you

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm [hunk 10]
```diff
@@ -777,6 +779,140 @@ sub get_all_Alleles_by_VariationFeature {
...
+  Caller     : general
+  Status     : Stable
+
+=cut
+
+sub get_all_PopulationGenotypes_by_VariationFeature {
+  my $self = shift;
+  my $vf = shift;
+  my $given_pop = shift;
+  my $vcf = shift;
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: Done with updated test in populationGenotypeAdaptor.t

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm (hunk 2 above)
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: I'm testing for an insertion which covers the elsif block.

File: modules/t/VCFCollection.t [hunk 11]
```diff
@@ -167,7 +168,54 @@ $coll->use_seq_region_synonyms(1);
...
+  [
+    'p:ESP6500:AA a:A f:1.0000 c:4406',
+    'p:ESP6500:AA a:C f:0.0000 c:0',
+    'p:ESP6500:EA a:A f:0.9999 c:8597',
+    'p:ESP6500:EA a:C f:0.0001 c:1'
+  ],
+  'get_all_Alleles_by_VariationFeature - freqs and counts ESP rs80359165'
+);
+
+my @population_genotypes = @{$coll->get_all_PopulationGenotypes_by_VariationFeature($vf)};
```
[the remark this answers is not in the document]
Comment (at7) [reply modules/t/VCFCollection.t]: added

[the remark this answers is not in the document]
Comment (at7) [reply modules/t/VCFCollection.t]: I just copied the structure of get_all_Alleles_by_VariationFeature.

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm (hunk 2 above)
Comment (at7) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: added also test for indel

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm (hunk 6 above)
[the remark this answers is not in the document]
Comment (helensch) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: Not that I can think of - was just checking as do not often use while(1) in API

File: modules/Bio/EnsEMBL/Variation/VCFCollection.pm (hunk 9 above)
[the remark this answers is not in the document]
Comment (helensch) [reply modules/Bio/EnsEMBL/Variation/VCFCollection.pm]: I put a ticket in JIRA

**Comment by at7** [comment]
Which PODs need fixing? I checked with podchecker and couldn't find any syntax problems.

**Comment by helensch** [comment]
podchecker passes but perldoc on modules display some formatting and boldness issues e.g.
VCFCollection.pm near synopsis and has_population. However these are not related to the PR and can create JIRA tickets for these.

**Comment by at7** [comment]
I can merge once you have approved the PR.

**Comment by helensch** [comment]
JIRA ticket create for perldoc display issues in 
VCFCollection.pm
PopulationGenotypeAdaptor.pm

**Comment by at7** [comment]
merged into master and release/99

**Pull request merged**